### PR TITLE
Adding request_path message

### PR DESCRIPTION
**Pull request by err508**
Implementing a message and a method for mocks/client to request paths (Client -> PFS).
Updating fee_info with web3 compliant serialize_bin.
Introducing a string fix for percentage_fee type. This is to be fixed.

### CODE REVIEW

File: raiden_libs/messages/request_paths.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+        target_address: Address,
+        value: int = 1,
+        num_paths: int = 1,
+        chain_id: int = 1,
+        nonce: int = 0,
+        signature: str = None,
+    ) -> None:
+        super().__init__()
+        assert is_address(token_network_address)
+        assert is_address(target_address)
```
Comment (pcppcp) [inline raiden_libs/messages/request_paths.py]: check all the parameters (ints should be unsigned + add max size check)
https://github.com/palango/raiden-libs/blob/19daf210185fc866e3eb3eb1902b27da1f222fa8/raiden_libs/messages/monitor_request.py#L32

File: raiden_libs/messages/request_paths.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+from raiden_libs.types import Address
+
+
+class RequestPaths(Message):
+    """ A message to request a path from PFS. It is sent from a raiden node to the PFS. """
+    def __init__(
+        self,
+        token_network_address: Address,
+        target_address: Address,
+        value: int = 1,
```
Comment (pcppcp) [inline raiden_libs/messages/request_paths.py]: do default values really make sense here? It'd be better to have all required fields enforced.

File: raiden_libs/messages/request_paths.py [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+            'num_paths': self.num_paths,
+            'chain_id': self.chain_id,
+            'nonce': self.nonce,
+            'signature': self.signature,
+        }
+
+    def serialize_bin(self):
+        """Return RequestPaths serialized to binary"""
+        order = '>20s32s32s8sd'
+        return struct.pack(
```
Comment (pcppcp) [inline raiden_libs/messages/request_paths.py]: please use `Web3` to pack the struct (it's easier to read as it uses solc data types)
https://github.com/palango/raiden-libs/blob/19daf210185fc866e3eb3eb1902b27da1f222fa8/raiden_libs/messages/balance_proof.py#L80

File: tests/test_messages.py [hunk 4]
```diff
@@ -40,6 +40,22 @@ def test_fee_info():
...
+        token_network_address='0x82dd0e0eA3E84D00Cc119c46Ee22060939E5D1FC',
+        target_address='0x82dd0e0eA3E84D00Cc119c46Ee22060939E5D1FC',
+        value=1000,
+        num_paths=1,
+        chain_id=1,
+        nonce=1,
+        signature='signature'
+    )
+
+    deserialized_message = Message.deserialize(message)
```
Comment (pcppcp) [inline tests/test_messages.py]: It'd be nice to add serialization test as well.
https://github.com/palango/raiden-libs/blob/19daf210185fc866e3eb3eb1902b27da1f222fa8/tests/test_messages.py#L79

File: raiden_libs/messages/__init__.py [hunk 5]
```diff
@@ -2,10 +2,12 @@
...
 from .fee_info import FeeInfo
+from .request_paths import RequestPaths
 
 __all__ = [
     'Message',
     'BalanceProof',
     'MonitorRequest',
-    'FeeInfo'
+    'FeeInfo',
+    'RequestPaths'
```
Comment (palango) [inline raiden_libs/messages/__init__.py]: We should specify some naming conventions here. There is the `MonitorRequest` which implies that it should be named `PathsRequest`.

File: raiden_libs/messages/json_schema.py [hunk 6]
```diff
@@ -111,6 +111,47 @@ def make_properties_required(schema):
...
+            'type': 'string',
+        },
+        'value': {
+            'type': 'integer',
+            'minimum': 1
+        },
+        'num_paths': {
+            'type': 'integer',
+            'minimum': 1
+
```
Comment (palango) [inline raiden_libs/messages/json_schema.py]: remove line

File: raiden_libs/messages/request_paths.py (hunk 2 above)
Comment (palango) [inline raiden_libs/messages/request_paths.py]: I'd set the value to 0 by default

File: raiden_libs/messages/request_paths.py [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+            'value': self.value,
+            'num_paths': self.num_paths,
+            'chain_id': self.chain_id,
+            'nonce': self.nonce,
+            'signature': self.signature,
+        }
+
+    def serialize_bin(self):
+        """Return RequestPaths serialized to binary"""
+        order = '>20s32s32s8sd'
```
Comment (palango) [inline raiden_libs/messages/request_paths.py]: the packing string needs to be adapted to the used values.

File: raiden_libs/messages/request_paths.py [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+            'target_address': self.target_address,
+            'value': self.value,
+            'num_paths': self.num_paths,
+            'chain_id': self.chain_id,
+            'nonce': self.nonce,
+            'signature': self.signature,
+        }
+
+    def serialize_bin(self):
+        """Return RequestPaths serialized to binary"""
```
Comment (palango) [inline raiden_libs/messages/request_paths.py]: s/Return/Returns

File: raiden_libs/messages/request_paths.py [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+
+    @classmethod
+    def deserialize(cls, data):
+        jsonschema.validate(data, REQUEST_PATHS_SCHEMA)
+        ret = cls(
+            token_network_address=data['token_network_address'],
+            target_address=data['target_address'],
+            value=int(data['value']),
+            num_paths=data['num_paths'],
+            chain_id=data['chain_id'],
```
Comment (palango) [inline raiden_libs/messages/request_paths.py]: this and chain_id should be ints as well

File: raiden_libs/test/mocks/client.py [hunk 10]
```diff
@@ -258,6 +259,21 @@ def get_fee_info(self, partner_address, **kwargs) -> FeeInfo:
         fi.signature = encode_hex(self.sign_data(fi.serialize_bin(), self.privkey))
         return fi
 
+    def request_paths(self, target_address, **kwargs) -> RequestPaths:
+        """Generate a RequestPaths for pathfinding-service .
```
Comment (palango) [inline raiden_libs/test/mocks/client.py]: s/Generate/Generates

File: raiden_libs/test/mocks/client.py [hunk 11]
```diff
@@ -258,6 +259,21 @@ def get_fee_info(self, partner_address, **kwargs) -> FeeInfo:
         fi.signature = encode_hex(self.sign_data(fi.serialize_bin(), self.privkey))
         return fi
 
+    def request_paths(self, target_address, **kwargs) -> RequestPaths:
```
Comment (palango) [inline raiden_libs/test/mocks/client.py]: Add a type annotation to `target_address`

File: raiden_libs/test/mocks/client.py [hunk 12]
```diff
@@ -258,6 +259,21 @@ def get_fee_info(self, partner_address, **kwargs) -> FeeInfo:
...
+        """Generate a RequestPaths for pathfinding-service .
+        Parameters:
+            token_network_address 	address
```
Comment (palango) [inline raiden_libs/test/mocks/client.py]: The parameters described here don't match the function signature. There should also be a space after the parameter name in the docstring. See https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html

File: raiden_libs/test/mocks/client.py [hunk 13]
```diff
@@ -258,6 +259,21 @@ def get_fee_info(self, partner_address, **kwargs) -> FeeInfo:
...
+            target_address - address of the transaction target, this method is agnostic
+            **kwargs - arguments to FeeInfo constructor
+        """
+        rp = RequestPaths(
```
Comment (palango) [inline raiden_libs/test/mocks/client.py]: use a better variable name here. `request`or similar

File: tests/test_messages.py [hunk 14]
```diff
@@ -54,7 +70,8 @@ def test_deserialize_with_required_type():
     deserialized_message = Message.deserialize(message, FeeInfo)
     assert isinstance(deserialized_message, FeeInfo)
 
-    # during deseriaisation the `message_type`is removed, add it back
+    # during deserialization the `message_type`is removed, add it back
```
Comment (palango) [inline tests/test_messages.py]: nice catch. please add a space after `message_type` as well and remove the newline.

File: raiden_libs/messages/request_paths.py (hunk 8 above)
Comment (err508) [reply raiden_libs/messages/request_paths.py]: I think  'return (a) RequestPaths (-message)' is correct here

File: raiden_libs/test/mocks/client.py (hunk 10 above)
Comment (err508) [reply raiden_libs/test/mocks/client.py]: same here

File: raiden_libs/messages/request_paths.py (hunk 2 above)
Comment (err508) [reply raiden_libs/messages/request_paths.py]: Agreed.

File: raiden_libs/messages/request_paths.py (hunk 8 above)
Comment (palango) [reply raiden_libs/messages/request_paths.py]: (This method) returns a RequestPaths (message)

Comment (err508) [reply raiden_libs/messages/request_paths.py]: (This method _should_) return a RequestPaths (-message). If it actually does, is to be checked by the reader of the code.

File: raiden_libs/messages/request_paths.py (hunk 3 above)
Comment (err508) [reply raiden_libs/messages/request_paths.py]: I think we should do a different PR to change it for all messages simultaneously, agreed?

File: raiden_libs/messages/request_paths.py (hunk 9 above)
Comment (err508) [reply raiden_libs/messages/request_paths.py]: Types are already enforced now.

**Comment by err508** [comment]
Yupp, I already started building pfs.in_request method.

**Comment by err508** [comment]
Merging still blocked for some reason.

**Comment by LefterisJP** [comment]
@pcppcp 's review blocks the merge here. Do you want me to dismiss it?

**Comment by LefterisJP** [comment]
@err508 make sure to check with @pcppcp if he has any additional concerns that can be handled in subsequent PRs. For now I merged this one as I don't want you to be blocked.

**Pull request merged**